### DISCUSSION TITLE
inconsistency in prometheus rate/increase/deriv/delta calculation

### DISCUSSION THREAD

**Discussion by puneets-ampere** [prometheus/prometheus#13900] [General]
Hi ,
While reviewing Prometheus-related [content online](https://stackoverflow.com/questions/54494394/do-i-understand-prometheuss-rate-vs-increase-functions-correctly), I've come to understand that the rate / deriv is calculated as (v2 - v1) / (t2 - t1), and the increase/delta is simply (v2 - v1).  

Now, considering i have 12 data points (of Gauge type) with a scrape interval of 1 minute, if i apply the rate(metric[2m]) / increase(metric[2m]) or deriv/delta functions, i notice some discrepancies in the results. 

Here is the example of data points which were collected by prometheus :
```
p1=10 @2024-04-06 13:33:48.666000
p2=10 @2024-04-06 13:34:48.666000
p3=10 @2024-04-06 13:35:48.666000
p4=10 @2024-04-06 13:36:48.666000
p5=20 @2024-04-06 13:37:48.666000
p6=20 @2024-04-06 13:38:48.666000
p7=20 @2024-04-06 13:39:48.666000
p8=20 @2024-04-06 13:40:48.666000
p9=30 @2024-04-06 13:41:48.666000
p10=30 @2024-04-06 13:42:48.666000
p11=30 @2024-04-06 13:43:48.666000
p12=30 @2024-04-06 13:44:48.666000
```

in that case if i use rate(metric[2m]), and increase(metric[2m]) , i expected the following:
```
p1: rate = NA,     increase=NA  ( not enough data points)   , 13:33:48.666000
p2: rate=  0    (10-10)/120    , increase = 0 (10-10)       , 13:34:48.666000   -   13:32:48.666000
p3:  rate=  0 (10-10)/120    , increase = 0 (10-10)    , 13:35:48.666000  -  13:33:48.666000 
p4:  rate=  0   (10-10)/120    , increase = 0 (10-10)   ,  13:36:48.666000  -  13:34:48.666000
p5: rate=0.0833 (20-10)/120 , increase=10 (20-10)  , 13:37:48.666000  -  13:35:48.666000 
p6:  rate=0.0833 (20-10)/120 , increase=10 (20-10)  , 13:38:48.666000 -  13:36:48.666000
p7:  rate=0 (20-20)/120 , increase=0 (20-20)  , 13:39:48.666000 - 13:37:48.666000
p8:  rate=0 (20-20)/120 , increase=0 (20-20)  , 13:40:48.666000 - 13:38:48.666000
p9:  rate=0.0833 (30-20)/120 , increase=10 (30-20) ,   13:41:48.666000 - 13:39:48.666000
p10:  rate=0.0833 (30-20)/120 , increase=10 (30-20) , 13:42:48.666000 - 13:40:48.666000
p11:  rate=0 (30-30)/120 , increase=0 (30-30)  , 12:43:48.666000  -  12:41:48.666000
p12:  rate=0 (30-30)/120 , increase=0 (30-30)  , 12:44:48.666000  -  12:42:48.666000
```

i tried visualizing the stored raw data , rate (2m) , increase(2m)  , deriv, delta  through the prometheus UI:
<img width="948" alt="increase" src="https://github.com/prometheus/prometheus/assets/154580421/8d3e03b8-3b1a-41c7-bce3-9d4fe7e08d77">

for point p5,p6 i expected a rate of 0.0833 and an increase of 10, but Prometheus displayed 0.16 and 20 respectively. Could you please share some insights into why prometheus could display 0.16 as rate/deriv and 20 as delta/increase ?

Regards,
P

**Comment by beorn7** [comment]
Maybe have a look at https://promlabs.com/blog/2021/01/29/how-exactly-does-promql-calculate-rates/ to get an better idea how the extrapolation really works?
